### PR TITLE
feat: add import JSON button for transformationEvent

### PR DESCRIPTION
**Pull request by huynguyen-hl**
## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents
[#4](https://github.com/gs-gs/pyx-acrs/issues/4)

Test data:
[aest-upload-vc1.json](https://github.com/uncefact/tests-untp/files/15433236/aest-upload-vc1.json)
[aest-upload-vc2.json](https://github.com/uncefact/tests-untp/files/15433237/aest-upload-vc2.json)

`app-config.json` example using for testing:

```json
{
  "name": "ACRS",
  "styles": {
    "primaryColor": "rgb(255, 207, 7)",
    "secondaryColor": "black",
    "tertiaryColor": "black"
  },
  "generalFeatures": [
    {
      "name": "General features",
      "type": "",
      "styles": {
        "primaryColor": "rgb(255, 207, 7)",
        "secondaryColor": "black",
        "tertiaryColor": "black"
      },
      "features": [
        {
          "name": "Conformity Credential",
          "id": "conformity_credential",
          "components": [
            {
              "name": "ConformityCredential",
              "type": "",
              "props": {
                "credentialRequestConfigs": [
                  {
                    "url": "http://example.com/deforestation-free-assessment",
                    "params": {},
                    "options": {
                      "headers": [],
                      "method": "POST"
                    },
                    "credentialName": "Deforestation Free Assessment",
                    "credentialPath": "/body/credentil",
                    "appOnly": "Farm"
                  },
                  {
                    "url": "http://example.com/ce",
                    "params": {},
                    "options": {
                      "headers": [],
                      "method": "GET"
                    },
                    "credentialName": "CE Marking",
                    "credentialPath": "",
                    "appOnly": "Farm"
                  },
                  {
                    "url": "http://example.com/conformity",
                    "params": {},
                    "options": {
                      "headers": [],
                      "method": "POST"
                    },
                    "credentialName": "Conformity",
                    "appOnly": "Feedlot"
                  }
                ],
                "storedCredentialsConfig": {
                  "url": "https://storage.example.com",
                  "params": {},
                  "options": {
                    "bucket": "bucket-stored-example"
                  },
                  "type": "s3"
                }
              }
            }
          ],
          "services": []
        }
      ]
    }
  ],
  "apps": [
    {
      "name": "Top Line Steel",
      "type": "producer",
      "assets": {
        "logo": "https://example.com/logo.png",
        "brandTitle": "Top Line Steel",
        "passportVC": "",
        "transactionEventVC": ""
      },
      "styles": {
        "primaryColor": "rgb(255, 207, 7)",
        "secondaryColor": "black",
        "tertiaryColor": "black"
      },
      "features": [
        {
          "name": "Upload VC",
          "components": [
            {
              "name": "RenderCheckList",
              "type": "EntryData",
              "props": {
                "requiredFields": ["steelID"],
                "checkBoxLabel": "Imported valid json:",
                "nestedComponents": [
                  {
                    "name": "ImportButton",
                    "type": "EntryData",
                    "props": {
                      "label": "Import JSON",
                      "style": { "margin": "40px auto", "paddingTop": "40px", "width": "80%" }
                    }
                  }
                ],
                "style": { "margin": "40px auto", "paddingTop": "40px", "width": "80%" }
              }
            },
            {
              "name": "CustomButton",
              "type": "Submit",
              "props": {
                "variant": "contained",
                "style": { "backgroundColor": "rgb(255, 207, 7)", "color": "black", "display": "block", "margin": "0 auto", "marginTop": "20px" }
              }
            }
          ],
          "services": [
            {
              "name": "mappingImportedData",
              "parameters": [
                {"requiredFields": ["steelID"]}
              ]
            },
            {
              "name": "processTransformationEvent",
              "parameters": [
                {
                  "epcisTransformationEvent": {
                    "context": ["https://dpp-json-ld.s3.ap-southeast-2.amazonaws.com/transformation-event-ld.json"],
                    "renderTemplate": [
                      { "template": "<p>Render epcis template</p>", "@type": "WebRenderingTemplate2022" }
                    ],
                    "type": ["TransformationEventCredential"],
                    "dlrIdentificationKeyType": "gtin",
                    "dlrLinkTitle": "EPCIS transformation event VC",
                    "dlrVerificationPage": "https://localhost:3000/verify",
                    "dlrQualifierPath": ""
                  },
                  "dpp": {
                    "context": ["https://dpp-json-ld.s3.ap-southeast-2.amazonaws.com/dppld.json"],
                    "renderTemplate": [
                      { "template": "<p>Render dpp template</p>", "@type": "WebRenderingTemplate2022" }
                    ],
                    "type": ["DigitalProductPassport"],
                    "dlrIdentificationKeyType": "gtin",
                    "dlrLinkTitle": "Digital Product Passport",
                    "dlrVerificationPage": "http://localhost:3000/verify",
                    "dlrQualifierPath": ""
                  },
                  "vckit": {
                    "vckitAPIUrl": "http://localhost:3332",
                    "issuer": "did:web:708f-27-75-129-38.ngrok-free.app"
                  },
                  "dlr": {
                    "dlrAPIUrl": "http://localhost:8080",
                    "dlrAPIKey": "5555555555555"
                  },
                  "storage": {
                    "storageAPIUrl": "https://storage.agtrace.showthething.com",
                    "bucket": "agtrace-test-verifiable-credentials"
                  },
                  "productTransformation": {
                    "inputItems": [{ "quantity": 1, "uom": "tons", "productClass": "Steel" }],
                    "outputItems": [
                      {
                        "productID": "0109359502000034",
                        "productClass": "Welding steel",
                        "quantity": 300,
                        "weight": 200,
                        "uom": "kilogram",
                        "description": "Welding steel is the process of joining steel pieces together using heat and pressure to create a strong, durable bond"
                      },
                      {
                        "productID": "0109359502000010",
                        "productClass": "Incorporating steel",
                        "quantity": 300,
                        "weight": 100,
                        "uom": "kilogram",
                        "description": "Incorporating steel involves integrating steel materials into various products, structures, or systems"
                      }
                    ]
                  },
                  "identifierKeyPaths": ["steelID"]
                }
              ]
            }
          ]
        },
        {
          "name": "Issue DPP",
          "id": "produce_product",
          "components": [
            {
              "name": "JsonForm",
              "type": "EntryData",
              "props": {
                "schema": {
                  "type": "object",
                  "properties": {
                    "product": {
                      "type": "object",
                      "properties": {
                        "modelName": {
                          "type": "string"
                        },
                        "identifier": {
                          "type": "string",
                          "title": "Identifier (GS1 Barcode)"
                        },
                        "traceabilityInformation": {
                          "type": "array",
                          "title": "Traceability Information",
                          "items": {
                            "type": "object",
                            "properties": {
                              "EventReference": {
                                "type": "string"
                              },
                              "EventType": {
                                "type": "string"
                              }
                            }
                          }
                        }
                      }
                    },
                    "issuedBy": {
                      "type": "object",
                      "properties": {
                        "name": {
                          "type": "string"
                        },
                        "identifiers": {
                          "type": "array",
                          "items": {
                            "type": "object",
                            "properties": {
                              "identifierURI": {
                                "type": "string"
                              },
                              "identiferValue": {
                                "type": "string"
                              }
                            }
                          }
                        }
                      }
                    },
                    "sustainabilityScore": {
                      "type": "integer",
                      "title": "Sustainability Score"
                    },
                    "trustScore": {
                      "type": "integer",
                      "title": "Trust Score"
                    }
                  }
                },
                "initialData": {},
                "className": "json-form",
                "style": { "margin": "40px auto", "paddingTop": "40px", "width": "80%" }
              }
            },
            {
              "name": "CustomButton",
              "type": "Submit",
              "props": {
                "variant": "contained",
                "style": { "backgroundColor": "rgb(255, 207, 7)", "color": "black", "display": "block", "margin": "0 auto", "marginTop": "20px" }
              }
            }
          ],
          "services": [
            {
              "name": "processObjectEvent",
              "parameters": [
                {
                  "vckit": {
                    "vckitAPIUrl": "http://localhost:3332",
                    "issuer": "did:web:708f-27-75-129-38.ngrok-free.app"
                  },
                  "dpp": {
                    "context": ["https://dpp-json-ld.s3.ap-southeast-2.amazonaws.com/dppld.json"],
                    "renderTemplate": [
                      { "template": "<p>Render dpp template</p>", "@type": "WebRenderingTemplate2022" }
                    ],
                    "type": ["DigitalProductPassport"],
                    "dlrLinkTitle": "Steel Passport",
                    "dlrIdentificationKeyType": "gtin",
                    "dlrVerificationPage": "https://localhost:3000/verify"
                  },

                  "dlr": {
                    "dlrAPIUrl": "http://localhost:8080",
                    "dlrAPIKey": "5555555555555"
                  },
                  "storage": {
                    "storageAPIUrl": "https://storage.agtrace.showthething.com",
                    "bucket": "agtrace-test-verifiable-credentials"
                  },
                  "identifierKeyPaths": ["product", "identifier"]
                }
              ]
            }
          ]
        },
        {
          "name": "Sell",
          "id": "produce_product",
          "components": [
            {
              "name": "JsonForm",
              "type": "EntryData",
              "props": {
                "schema": {
                  "type": "object",
                  "properties": {
                    "sourceParty": {
                      "type": "object",
                      "properties": {
                        "partyID": {
                          "type": "string",
                          "enum": ["https://tomstoplinesteel.com/41161080146.json"]
                        },
                        "name": {
                          "type": "string",
                          "enum": ["Top Line Steel"]
                        }
                      }
                    },
                    "destinationParty": {
                      "type": "object",
                      "properties": {
                        "partyID": {
                          "type": "string",
                          "enum": ["https://sammyssteelprocessor.com/51161080100.json"]
                        },
                        "name": {
                          "type": "string",
                          "enum": ["Sammys Steel Processor"]
                        }
                      }
                    },
                    "transaction": {
                      "type": "object",
                      "properties": {
                        "type": {
                          "type": "string",
                          "enum": ["bol", "cert", "desadv", "inv", "po"]
                        },
                        "identifier": {
                          "type": "string",
                          "title": "Identifier (GS1 Barcode)"
                        },
                        "documentURL": {
                          "type": "string"
                        }
                      },
                      "required": ["identifier"]
                    },
                    "itemList": {
                      "type": "array",
                      "items": {
                        "type": "object",
                        "properties": {
                          "itemID": {
                            "type": "string"
                          },
                          "name": {
                            "type": "string"
                          }
                        }
                      }
                    },
                    "quantityList": {
                      "type": "array",
                      "items": {
                        "type": "object",
                        "properties": {
                          "productClass": {
                            "type": "string"
                          },
                          "quantity": {
                            "type": "string"
                          },
                          "uom": {
                            "type": "string"
                          }
                        }
                      }
                    }
                  }
                },
                "initialData": {},
                "className": "json-form",
                "style": { "margin": "40px auto", "paddingTop": "40px", "width": "80%" }
              }
            },
            {
              "name": "CustomButton",
              "type": "Submit",
              "props": {
                "variant": "contained",
                "style": { "backgroundColor": "rgb(255, 207, 7)", "color": "black", "display": "block", "margin": "0 auto", "marginTop": "20px" }
              }
            }
          ],
          "services": [
            {
              "name": "processTransactionEvent",
              "parameters": [
                {
                  "vckit": {
                    "vckitAPIUrl": "http://localhost:3332",
                    "issuer": "did:web:708f-27-75-129-38.ngrok-free.app"
                  },
                  "epcisTransactionEvent": {
                    "context": ["https://dpp-json-ld.s3.ap-southeast-2.amazonaws.com/dppld.json"],
                    "renderTemplate": [
                      { "template": "<p>Render dpp template</p>", "@type": "WebRenderingTemplate2022" }
                    ],
                    "type": ["TransactionEventCredential"],
                    "dlrLinkTitle": "Transaction Event",
                    "dlrIdentificationKeyType": "gtin",
                    "dlrVerificationPage": "https://localhost:3000/verify"
                  },
                  "dlr": {
                    "dlrAPIUrl": "http://localhost:8080",
                    "dlrAPIKey": "5555555555555"
                  },
                  "storage": {
                    "storageAPIUrl": "https://storage.agtrace.showthething.com",
                    "bucket": "agtrace-test-verifiable-credentials"
                  },
                  "identifierKeyPaths": ["transaction", "identifier"]
                }
              ]
            }
          ]
        }
      ]
    },
    {
      "name": "Conformity Credential",
      "type": "",
      "assets": {
        "logo": "",
        "brandTitle": "Conformity Credential",
        "passportVC": "",
        "transactionEventVC": ""
      },
      "styles": {
        "primaryColor": "rgb(255, 207, 7)",
        "secondaryColor": "black",
        "tertiaryColor": "black"
      },
      "features": [
        {
          "name": "Conformity Credential",
          "id": "conformity_credential",
          "components": [
            {
              "name": "CustomButton",
              "type": "Submit",
              "props": {
                "style": { "textAlign": "center", "paddingTop": "40px" },
                "label": "Request Conformity Credential"
              }
            }
          ],
          "services": [
            {
              "name": "conformityCredentialService",
              "parameters": [
                {
                  "credentialRequestConfig": [
                    {
                      "url": "http://example.com/getConformityCredential",
                      "params": {
                        "id": "conformityCredential"
                      },
                      "options": {
                        "headers": [],
                        "method": "POST"
                      }
                    }
                  ],
                  "uploadCredentialConfig": {
                    "url": "https://storage.example.com",
                    "params": {},
                    "options": {
                      "bucket": "bucket-stored-example"
                    },
                    "type": "s3"
                  }
                }
              ]
            }
          ]
        }
      ]
    }
  ],
  "identifyProvider": {
    "type": "gs1",
    "url": "http://localhost:3333/products"
  },
  "defaultVerificationServiceLink": {
    "title": "Default Verification Service",
    "context": "Default Verification Service",
    "type": "application/json",
    "href": "http://localhost:3332/agent/routeVerificationCredential",
    "hreflang": ["en"]
  }
}
```

## Mobile & Desktop Screenshots/Recordings

Screenshot 2024-05-23 at 5 53 02 PM
Screenshot 2024-05-23 at 5 58 57 PM
Screenshot 2024-05-23 at 5 59 35 PM

## Added tests?

- [x] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [x] 📕 storybook
- [ ] 🙅 no documentation needed

### CODE REVIEW

**Comment by ldhyen99** [comment]
Hi @huynguyen-hl, can you add `nestedComponents` to `app-config.example.json` for easy to follow? 

```json
...
 "nestedComponents": [
  {
    "name": "ImportButton",
    "type": "EntryData",
    "props": {
      "label": "Import JSON",
      "style": { "margin": "40px auto", "paddingTop": "40px", "width": "80%" }
    }
  }
],
...
```

File: packages/components/src/components/RenderCheckList/RenderCheckList.tsx [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+              if (!requiredFieldValue) {
+                return null;
+              }
+              
+              const validData = { label: requiredFieldValue, value: data };
+              setRenderCheckBoxList((prev) => [...prev, validData]);
+            };
+
+            break;
+            // case ComponentType.Submit:
```
Comment (ldhyen99) [inline packages/components/src/components/RenderCheckList/RenderCheckList.tsx:86]: Should we delete unused code?

**Comment by ldhyen99** [comment]
Hey @huynguyen-hl, I have a problem when using the `app-config.json`, the error is on `generateFeatures` field. In the `src/pages/Home.tsx`, we're using `generalFeatures` to map the value.
image

**Comment by ldhyen99** [comment]
I also found some issue in UI:
`/top-line-steel/upload-vc`
image

**Comment by huynguyen-hl** [comment]
> I also found some issue in UI: `/top-line-steel/upload-vc` image
> 
> `/top-line-steel/issue-dpp` image

@ldhyen99, You can use the latest `app-config.json` content above. It contains the `CustomButton` styles, so I think it can fix the `Submit` style problem that you encountered.

File: packages/mock-app/src/components/GenericFeature/GenericFeature.tsx [hunk 2]
```diff
@@ -141,7 +141,17 @@ export const GenericFeature: React.FC<IGenericFeatureProps> = ({ components, ser
...
           case ComponentType.Submit:
             props.onClick = async () => {
-              await executeServices(services, state);
+              try {
+                const result = await executeServices(services, state);
+                if (!result)  {
+                  return toastMessage({ status: Status.error, message: 'Something went wrong' });  
+                }
+
+                toastMessage({ status: Status.success, message: 'Submit success' });
```
Comment (ldhyen99) [inline packages/mock-app/src/components/GenericFeature/GenericFeature.tsx:150]: Should we hardcode the `Submit` message here? What if the button is Download or ScanQR button?

**Review by ldhyen99 — DISMISSED** [review]

**Review by namhoang1604 — APPROVED** [review]

**Pull request merged**